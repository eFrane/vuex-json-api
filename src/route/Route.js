export class Route {
  /**
   *
   * @param {String} module
   * @param {String} action
   * @param {String} url
   * @param {Array} parameters
   */
  constructor (module, action, url, parameters) {
    if (typeof module !== 'string') {
      throw new Error('Module name must be string')
    }

    if (!this.isValidAction(action)) {
      throw new Error('Action must be valid resourceful verb')
    }

    if (typeof url !== 'string') {
      throw new Error('URL must be string')
    }

    if (parameters.constructor !== Array) {
      throw new Error('Parameters must be array')
    }

    this.module = module
    this.action = action.toLowerCase()
    this.url = url
    this.parameters = parameters
  }

  get supportedResourcefulVerbs () {
    return [
      'list',
      'get',
      'create',
      'replace',
      'update',
      'delete'
    ]
  }

  /**
   * Prepare an url, replacing provided parameters
   *
   * @param parameters
   * @returns {string}
   */
  prepare (parameters) {
    let url = this.url

    for (const param in parameters) {
      if (Object.prototype.hasOwnProperty.call(parameters, param) &&
        this.hasParameter(param)) {
        url = url.replace('{' + param + '}', parameters[param])
        delete parameters[param]
      }
    }

    return url
  }

  /**
   * Check if a parameter is allowed
   *
   * @param parameter
   * @returns {boolean}
   */
  hasParameter (parameter) {
    return this.parameters.filter((checkParam) => {
      return parameter === checkParam
    }).length > 0
  }

  isAbsolute () {
    return this.url.substring(0, 4) === 'http'
  }

  /**
   *
   * @param {String} action
   */
  isValidAction (action) {
    return typeof action === 'string' && this.supportedResourcefulVerbs.indexOf(action.toLowerCase()) > -1
  }
}
