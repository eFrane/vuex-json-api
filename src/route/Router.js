/**
 * Basic router implementation for the ResourcefulApi.
 *
 * Automagically creating api bound modules builds on
 * an understanding of the available routes. To
 * easily instantiate a Store bound to an endpoint,
 * route information for that endpoint must be provided.
 *
 * Since every endpoint is implemented differently and
 * the choice where this route information comes from
 * should be left to the endpoint developer, this
 * library only assumes that route loading is usually an
 * asynchronous process which eventually returns and
 * has a set of keyed-by-name `Route` objects in
 * `this.routes`.
 */
export class Router {
  constructor () {
    this.routes = {}
  }

  getRoutes () {
    return this.routes
  }

  async updateRoutes () {
    return this
  }

  /**
   *
   * @param {Route} route
   */
  addRoute (route) {
    if (!Object.prototype.hasOwnProperty.call(this.routes, route.module)) {
      this.routes[route.module] = {}
    }

    this.routes[route.module][route.action] = route
  }
}
