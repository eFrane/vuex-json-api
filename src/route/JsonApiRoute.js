import { Route } from './Route'

function getRequiredAttribute (routeResource, attributeName) {
  if (!Object.prototype.hasOwnProperty.call(routeResource.attributes, attributeName)) {
    throw new Error(`Missing required resource attribute: ${attributeName}`)
  }

  return routeResource.attributes[attributeName]
}

export class JsonApiRoute extends Route {
  /**
   * Create a route from a Json:Api resource of the VuexJsonApiRoute-type:
   *
   * ``` json
   * {
   *   "type": "VuexJsonApiRoute",
   *   "attributes": {
   *     "module": "themodule",
   *     "method": "thehttpmethod"
   *     "url": "The URL of the Resource, can be relative to baseURL",
   *     "parameters": ["Parameters", "for", "the", "above", url"]
   *   }
   * }
   * ```
   *
   * @param routeResource
   */
  constructor (routeResource) {
    if (!Object.prototype.hasOwnProperty.call(routeResource, 'attributes') ||
      !Object.prototype.hasOwnProperty.call(routeResource, 'type')) {
      throw new Error('Expected json:api-like object structure')
    }

    if (routeResource.type.toLowerCase() !== 'vuexjsonapiroute') {
      throw new Error('Invalid resource type')
    }

    const module = getRequiredAttribute(routeResource, 'module')
    const method = getRequiredAttribute(routeResource, 'method')
    const url = getRequiredAttribute(routeResource, 'url')
    const parameters = routeResource.attributes.parameters || []

    super(module, method, url, parameters)
  }
}
