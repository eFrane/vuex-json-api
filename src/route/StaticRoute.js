import { Route } from './Route'

/**
 * Parses a static route data object into a `Route`
 *
 * ``` json
 * [
 *   {
 *     "module": "MyApiModule",
 *     "action": "GET",
 *     "url": "/api/my-api-module/{id}",
 *     "parameters": ["id"]
 *   }
 * ]
 * ```
 */
export class StaticRoute extends Route {
  constructor (routeData) {
    const module = routeData.module
    const action = routeData.action
    const url = routeData.url
    const parameters = routeData.parameters || []

    super(module, action, url, parameters)
  }
}
