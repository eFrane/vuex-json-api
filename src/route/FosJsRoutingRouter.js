import { Route } from './Route'
import { Router } from './Router'

/**
 * Pluggable api router if you're using Symfony and the FosJsRouting Bundle
 *
 * @see https://github.com/isychev/fos-routing/
 * @see https://github.com/FriendsOfSymfony/FOSJsRoutingBundle
 */
export class FosJsRoutingRouter extends Router {
  constructor (fosRouter, options = {}) {
    super()

    this.fosRouter = fosRouter
    this.routeExtractionOptions = Object.assign({
      prefix: '',
      resourceNamePattern: '[a-b]+',
      routePatterns: {
        resourceRoute: '{prefix}_{resourceName}_{action}'
      },
      actionPatterns: {
        list: 'list',
        get: 'get',
        create: 'create',
        replace: 'replace',
        update: 'update',
        delete: 'delete'
      }
    }, options)
  }

  async updateRoutes () {
    // TODO: extract routes from fosRouter based on naming scheme or something
    this.fosRouter.map(this.fosRouter.getRoutes(), name => {
      if (this.matchRouteName(name)) {
        const route = Route.fromPojo()
        this.addRoute(route)
      }
    })
  }

  matchRouteName (name) {

  }

  prepareResourceMatchRegex () {
    const pattern = this.routeExtractionOptions.routePatterns.resourceRoute
      .replace(
        '{prefix}',
        this.routeExtractionOptions.prefix
      )
      .replace(
        '{resourceName}',
        `(?<resourceName>${this.routeExtractionOptions.resourceNamePattern})`
      )
      .replace(
        '{action}',
        `(?<action>${Object.values(this.routeExtractionOptions.actionPatterns).join('|')})`
      )

    return new RegExp(pattern)
  }
}
