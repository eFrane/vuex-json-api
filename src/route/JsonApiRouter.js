import { Api } from '../api/Api'
import { Router } from './Router'
import { JsonApiRoute } from './JsonApiRoute'

/**
 * A simple router implementation querying a json:api endpoint which delivers all
 * necessary route information
 *
 * The endpoint must return a list of Route objects:
 *
 * ```json
 * {
 *     "data": [
 *         {
 *             "type": "Route",
 *             "id": "api.route.list",
 *             "attributes": {
 *                 "parameters": [],
 *                 "url": "api/route",
 *                 "method": "list"
 *             }
 *        },
 *        {
 *            "type": "Route",
 *            "id": "api.route.get",
 *            "attributes": {
 *                "parameters": [
 *                    "id"
 *                ],
 *                "url": "api/route/{id}",
 *                "method": "get"
 *            }
 *        }
 *     ]
 * }
 * ```
 *
 * To initialize this router, simply `new` it with the desired fetch path,
 * e.g. `new JsonApiRouter('/api/route')`.
 */
export class JsonApiRouter extends Router {
  constructor (fetchPath) {
    super()
    this.fetchPath = fetchPath
  }

  /**
   * Update routes
   *
   * @returns {*}
   */
  async updateRoutes () {
    const api = new Api()
    return api.get(this.fetchPath)
      .then(({ data }) => {
        console.time('router_setup')
        for (const idx in data.route) {
          if (Object.prototype.hasOwnProperty.call(data.route, idx)) {
            const routeResource = data.route[idx]

            this.addRoute(new JsonApiRoute(routeResource))
          }
        }
        console.timeEnd('router_setup')

        // returning self to make working with this in chained promises easier
        return this
      })
  }
}
