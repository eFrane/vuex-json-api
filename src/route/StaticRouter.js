import { Router } from './Router'
import { StaticRoute } from './StaticRoute'

/**
 * Static router
 *
 * It may be desirable to not do an additional request
 * for getting the api routes and instead bake them into
 * the code. To this avail, the `StaticRouter` can be
 * initialized with a POJO of { id, url, parameters }
 * whereas parameters are parts of the url which can be replaced.
 *
 *
 */
export class StaticRouter extends Router {
  /**
   * Requires an array of route data objects.
   *
   * @see StaticRoute
   * @param {Array} routes
   */
  constructor (routes) {
    super()

    for (const idx in routes) {
      const routeData = routes[idx]
      this.addRoute(new StaticRoute(routeData))
    }
  }
}
