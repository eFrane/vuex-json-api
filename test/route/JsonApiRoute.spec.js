import { Route } from '@/route/Route'
import { JsonApiRoute } from '@/route/JsonApiRoute'

describe('JsonApiRoute', () => {
  it('processes json api responses', () => {
    const resource = {
      type: 'vuexjsonapiroute',
      id: 2,
      attributes: {
        module: 'tag',
        method: 'get',
        url: '/tag/{id}',
        parameters: [
          'id'
        ]
      }
    }

    const route = new JsonApiRoute(resource)

    expect(route).toBeInstanceOf(Route)
    expect(route.module).toStrictEqual('tag')
    expect(route.parameters).toStrictEqual(['id'])
  })

  it('fails on wrong type', () => {
    const wrongTypeTest = function () {
      const resource = {
        type: 'routeforvuexjsonapi',
        id: 2,
        attributes: {
          module: 'tag',
          method: 'get',
          url: '/tag/{id}',
          parameters: [
            'id'
          ]
        }
      }

      /* eslint-disable no-new */
      new JsonApiRoute(resource)
    }

    expect(wrongTypeTest).toThrowErrorMatchingSnapshot()
  })
})
