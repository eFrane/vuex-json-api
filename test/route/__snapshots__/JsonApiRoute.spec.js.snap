// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonApiRoute fails on wrong type 1`] = `"Invalid resource type"`;
